File: algo/mergesort.py
sorted_data = [0]*8
numbers = [7,6,5,8,3,5,9,1]

def merge(data, m, middle, n):
    i = m
    j = middle + 1
    k = m
    # 작은 순서대로 배열에 삽입
    while i <= middle and j <= n:
        if data[i] <= data[j]:
            sorted_data[k] = data[i]
            i += 1
        else:
            sorted_data[k] = data[j]
            j +=1
        k += 1
    #남은 데이터도 삽입
    if i > middle:
        for t in range(j,n+1):
            sorted_data[k] = data[t]
            k += 1
    else:
        for t in range(i,middle+1):
            sorted_data[k] = data[t]
            k += 1

    for t in range(m,n+1):
        data[t] = sorted_data[t];

def mergeSort(data, m, n):
    if m < n:
        middle = (m + n) // 2
        mergeSort(data, m, middle)
        mergeSort(data, middle + 1, n)
        merge(data, m, middle, n)



mergeSort(numbers, 0, 7)
print(numbers)

File: etc/sort/merge_sort.cpp
/*
병합정렬 (Merge Sort)
	분할정복 알고리즘
	정확히 반절씩 나눈다는 점에서 최악의 경우에도 O(N*logN)
	기존의 데이터를 담을 추가적인 배열 공간이 필요
	O(N*logN)
*/

#include <iostream>

using namespace std;

int sorted[10];

void show(int* numbers) {
	for (int i = 0; i < 10; i++)
	{
		cout << numbers[i] << " ";
	}
}

void merge(int *numbers, int m, int middle, int n) {
	int i = m;
	int j = middle + 1;
	int k = m;

	while (i <= middle && j <= n)
	{
		if (numbers[i] < numbers[j])
		{
			sorted[k] = numbers[i];
			i++;
		}
		else
		{
			sorted[k] = numbers[j];
			j++;
		}
		k++;
	}

	if (i>middle)
	{
		for (int a = j; a <= n ; a++)
		{
			sorted[k] = numbers[a];
			k++;
		}
	}
	else
	{
		for (int a = i; a <= middle; a++)
		{
			sorted[k] = numbers[a];
			k++;
		}
	}

	for (int a = m; a <= n; a++)
	{
		numbers[a] = sorted[a];
	}
}

void merge_sort(int* numbers, int m, int n) {
	if (m < n)
	{
		int middle = (m + n) / 2;
		merge_sort(numbers, m, middle);
		merge_sort(numbers, middle + 1, n);
		merge(numbers, m, middle, n);
	}
}

int main(void) {
	int numbers[10] = { 1,3,4,5,2,6,8,9,7,10 };

	merge_sort(numbers, 0, 9);
	show(numbers);
	return 0;
}

